tracer/src/Datadog.Trace/Util/AsyncManualResetEvent.cs

```diff
@@ -0,0 +1,121 @@
+// <copyright file="AsyncManualResetEvent.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+#nullable enable
+
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Datadog.Trace.Util;
+
+// Based on: https://github.com/StephenCleary/AsyncEx/blob/master/src/Nito.AsyncEx.Coordination/AsyncManualResetEvent.cs
+
+internal class AsyncManualResetEvent
+{
+    private readonly object _mutex;
+    private TaskCompletionSource<object?> _tcs;
+
+    public AsyncManualResetEvent(bool set)
+    {
+        _mutex = new object();
+        _tcs = new TaskCompletionSource<object?>(TaskCreationOptions.RunContinuationsAsynchronously);
+        if (set)
+        {
+            _tcs.TrySetResult(null);
+        }
+    }
+
+    public AsyncManualResetEvent()
+        : this(false)
+    {
+    }
+
+    public bool IsSet
+    {
+        get
+        {
+            lock (_mutex)
+            {
+                return _tcs.Task.IsCompleted;
+            }
+        }
+    }
+
+    public Task WaitAsync()
+    {
+        lock (_mutex)
+        {
+            return _tcs.Task;
+        }
+    }
+
+    public Task WaitAsync(int millisecondTimeout)
+    {
+        var waitTask = WaitAsync();
+
+        return waitTask.IsCompleted ? waitTask : InternalWaitAsync(waitTask, millisecondTimeout);
+
+        static async Task InternalWaitAsync(Task task, int timeout)
+        {
+            using var delayCancellation = new CancellationTokenSource();
+            var completedTask = await Task.WhenAny(task, Task.Delay(timeout, delayCancellation.Token)).ConfigureAwait(false);
+            if (completedTask == task)
+            {
+                delayCancellation.Cancel();
+                await task.ConfigureAwait(false);
+            }
+        }
+    }
+
+    public Task WaitAsync(CancellationToken cancellationToken)
+    {
+        var waitTask = WaitAsync();
+        if (waitTask.IsCompleted)
+        {
+            return waitTask;
+        }
+
+        if (!cancellationToken.CanBeCanceled)
+        {
+            return waitTask;
+        }
+
+        if (cancellationToken.IsCancellationRequested)
+        {
+            return Task.FromCanceled(cancellationToken);
+        }
+
+        return DoWaitAsync(waitTask, cancellationToken);
```

Comment:
I think you can simplify this for .NET6+
```suggestion
#if NET6_0_OR_GREATER
        return waitTask.WaitAsync(cancellationToken);
#else
        return DoWaitAsync(waitTask, cancellationToken);
#endif        
```

Reply:
done in https://github.com/DataDog/dd-trace-dotnet/pull/3609/commits/18bfc35e3ce644ce86d4f155a7ffde9667c3a1d8